nixos/modules/programs/fish.nix

```diff
@@ -169,18 +178,75 @@ in
       end
     '';
 
+    programs.fish.interactiveShellInit = ''
+      # add completions generated by NixOS to $fish_complete_path
+      begin
+        set -l joined (string join " " $fish_complete_path)
+        set -l prev_joined (string replace --regex "[^\s]*generated_completions.*" "" $joined)
+        set -l post_joined (string replace $prev_joined "" $joined)
+        set -l prev (string split " " (string trim $prev_joined))
+        set -l post (string split " " (string trim $post_joined))
+        set fish_complete_path $prev "/etc/fish/generated_completions" $post
+      end
+    '';
+
+    environment.etc."fish/generated_completions".source =
+      let
+        # paths later in the list will overwrite those already linked
+        destructiveSymlinkJoin =
+          args_@{ name
+              , paths
+              , preferLocalBuild ? true
+              , allowSubstitutes ? false
+              , postBuild ? ""
+              , ...
+              }:
+          let
+            args = removeAttrs args_ [ "name" "postBuild" ]
+              // { inherit preferLocalBuild allowSubstitutes; }; # pass the defaults
+          in pkgs.runCommand name args
+            ''
+              mkdir -p $out
+              for i in $paths; do
```

Comment:
`$paths` should be quoted, see https://github.com/koalaman/shellcheck/wiki/Sc2086

Reply:
We do want splitting on spaces here though, since that is the format the nix attribute `paths`of type list is converted to.